src/mono/wasm/runtime/loader/config.ts

```diff
@@ -83,25 +93,87 @@ function deep_merge_resources(target: ResourceGroups, source: ResourceGroups): R
         providedResources.extensions = { ...(target.extensions || {}), ...(providedResources.extensions || {}) };
     }
     if (providedResources.vfs !== undefined) {
-        providedResources.vfs = { ...(target.vfs || {}), ...(providedResources.vfs || {}) };
+        providedResources.vfs = deep_merge_dict(target.vfs || {}, providedResources.vfs || {});
     }
     return Object.assign(target, providedResources);
 }
 
+function deep_merge_dict(target: { [key: string]: ResourceList }, source: { [key: string]: ResourceList }) {
+    // no need to merge the same object
+    if (target === source) return target;
+
+    for (const key in source) {
+        target[key] = { ...target[key], ...source[key] };
+    }
+    return target;
+}
+
 // NOTE: this is called before setRuntimeGlobals
 export function normalizeConfig() {
     // normalize
     const config = loaderHelpers.config;
 
     config.environmentVariables = config.environmentVariables || {};
-    config.assets = config.assets || [];
     config.runtimeOptions = config.runtimeOptions || [];
     config.resources = config.resources || {
         assembly: {},
         jsModuleNative: {},
+        jsModuleWorker: {},
         jsModuleRuntime: {},
-        wasmNative: {}
+        wasmNative: {},
+        vfs: {},
+        satelliteResources: {},
     };
+
+    if (config.assets) {
+        mono_log_debug("assets are deprecated, use resources instead");
```

Comment:
it would probably be better to phrase this like "config.assets is deprecated, use config.resources instead" but not terribly important